### ISSUE TITLE
pipeline.yml

### ISSUE THREAD

**Issue by Rebonegcabo** [Rebonegcabo/.github-workflows-#1]
%%writefile .github/workflows/pipeline.yml
name: Tourism Project Pipeline

on:
  push:
    branches:
      - main  # Automatically triggers on push to the main branch

jobs:

  register-dataset:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3
      - name: Install Dependencies
        run: pip install -r tourism_project/requirements_workflow.txt
      - name: Create data directory
        run: mkdir -p tourism_project/data
      - name: Copy dataset
        run: cp /content/tourism.csv tourism_project/data/
      - name: Upload Dataset to Hugging Face Hub
        env:
          HF_TOKEN: ${{ secrets.HF_TOKEN }}
        run: |
          pip install huggingface_hub
          python -c "from huggingface_hub import HfApi; api = HfApi(); api.upload_file(path_or_fileobj='tourism_project/data/tourism.csv', path_in_repo='tourism.csv', repo_id='your-hf-username/your-hf-dataset-repo', repo_type='dataset')"

  data-prep:
    needs: register-dataset
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3
      - name: Install Dependencies
        run: pip install -r tourism_project/requirements_workflow.txt
      - name: Download Dataset from Hugging Face Hub
        env:
          HF_TOKEN: ${{ secrets.HF_TOKEN }}
        run: |
          pip install huggingface_hub
          python -c "from huggingface_hub import hf_hub_download; hf_hub_download(repo_id='your-hf-username/your-hf-dataset-repo', filename='tourism.csv', local_dir='.')"
          mkdir -p tourism_project/data
          mv tourism.csv tourism_project/data/
      - name: Run Data Preparation
        env:
          HF_TOKEN: ${{ secrets.HF_TOKEN }}
        run: |
          # Execute data preparation steps. This would typically be a script.
          # For this example, we'll simulate it, but in a real scenario, you'd run a python script here.
          python -c "
import pandas as pd
from sklearn.preprocessing import OneHotEncoder
from sklearn.model_selection import train_test_split
import joblib

df = pd.read_csv('tourism_project/data/tourism.csv')
categorical_cols = df.select_dtypes(include='object').columns
encoder = OneHotEncoder(sparse_output=False, handle_unknown='ignore')
encoded_features = encoder.fit_transform(df[categorical_cols])
encoded_df = pd.DataFrame(encoded_features, columns=encoder.get_feature_names_out(categorical_cols), index=df.index)
df_encoded = pd.concat([df.drop(columns=categorical_cols), encoded_df], axis=1)
X = df_encoded.drop('ProdTaken', axis=1)
y = df_encoded['ProdTaken']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)
# Save processed data for model training step
X_train.to_csv('tourism_project/X_train.csv', index=False)
y_train.to_csv('tourism_project/y_train.csv', index=False)
X_test.to_csv('tourism_project/X_test.csv', index=False)
y_test.to_csv('tourism_project/y_test.csv', index=False)
encoder_path = 'tourism_project/encoder.joblib'
joblib.dump(encoder, encoder_path)
          "
      - name: Upload processed data and encoder to Hugging Face Hub
        env:
          HF_TOKEN: ${{ secrets.HF_TOKEN }}
        run: |
          pip install huggingface_hub joblib
          python -c "from huggingface_hub import HfApi; api = HfApi(); api.upload_file(path_or_fileobj='tourism_project/X_train.csv', path_in_repo='X_train.csv', repo_id='your-hf-username/your-hf-model-repo'); api.upload_file(path_or_fileobj='tourism_project/y_train.csv', path_in_repo='y_train.csv', repo_id='your-hf-username/your-hf-model-repo'); api.upload_file(path_or_fileobj='tourism_project/X_test.csv', path_in_repo='X_test.csv', repo_id='your-hf-username/your-hf-model-repo'); api.upload_file(path_or_fileobj='tourism_project/y_test.csv', path_in_repo='y_test.csv', repo_id='your-hf-username/your-hf-model-repo'); api.upload_file(path_or_fileobj='tourism_project/encoder.joblib', path_in_repo='encoder.joblib', repo_id='your-hf-username/your-hf-model-repo')"

  model-training:
    needs: data-prep
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3
      - name: Install Dependencies
        run: pip install -r tourism_project/requirements_workflow.txt
      - name: Download processed data from Hugging Face Hub
        env:
          HF_TOKEN: ${{ secrets.HF_TOKEN }}
        run: |
          pip install huggingface_hub
          python -c "from huggingface_hub import hf_hub_download; hf_hub_download(repo_id='your-hf-username/your-hf-model-repo', filename='X_train.csv', local_dir='tourism_project'); hf_hub_download(repo_id='your-hf-username/your-hf-model-repo', filename='y_train.csv', local_dir='tourism_project')"
      - name: Start MLflow Server
        run: |
          nohup mlflow ui --host 0.0.0.0 --port 5000 &  # Run MLflow UI in the background
          sleep 5  # Wait for a moment to let the server starts
      - name: Model Building
        env:
          HF_TOKEN: ${{ secrets.HF_TOKEN }}
        run: |
          python -c "
import pandas as pd
import joblib
from sklearn.ensemble import RandomForestClassifier
import mlflow

mlflow.set_tracking_uri('http://127.0.0.1:5000')
mlflow.set_experiment('Tourism_Prediction')

X_train = pd.read_csv('tourism_project/X_train.csv')
y_train = pd.read_csv('tourism_project/y_train.csv').squeeze()

with mlflow.start_run():
    rf_model = RandomForestClassifier(n_estimators=100, random_state=42)
    rf_model.fit(X_train, y_train)

    # Log model parameters
    mlflow.log_param('n_estimators', 100)
    mlflow.log_param('random_state', 42)

    # Log the model
    mlflow.sklearn.log_model(rf_model, 'random_forest_model')

    # Save the model locally as well for deployment
    model_path = 'tourism_project/rf_model.joblib'
    joblib.dump(rf_model, model_path)

    # Upload the model to Hugging Face Hub
    from huggingface_hub import HfApi
    api = HfApi(); api.upload_file(path_or_fileobj=model_path, path_in_repo='rf_model.joblib', repo_id='your-hf-username/your-hf-model-repo')
          "

  deploy-hosting:
    runs-on: ubuntu-latest
    needs: [model-training,data-prep,register-dataset]
    steps:
      - uses: actions/checkout@v3
      - name: Install Dependencies
        run: pip install -r tourism_project/requirements_workflow.txt
      - name: Download Model from Hugging Face Hub
        env:
          HF_TOKEN: ${{ secrets.HF_TOKEN }}
        run: |
          pip install huggingface_hub
          python -c "from huggingface_hub import hf_hub_download; hf_hub_download(repo_id='your-hf-username/your-hf-model-repo', filename='rf_model.joblib', local_dir='tourism_project/deployment/')"
          python -c "from huggingface_hub import hf_hub_download; hf_hub_download(repo_id='your-hf-username/your-hf-model-repo', filename='encoder.joblib', local_dir='tourism_project/deployment/')"
          python -c "from huggingface_hub import hf_hub_download; hf_hub_download(repo_id='your-hf-username/your-hf-dataset-repo', filename='tourism.csv', local_dir='.')"
          mkdir -p tourism_project/data
          mv tourism.csv tourism_project/data/

      - name: Push files to Frontend Hugging Face Space
        env:
          HF_TOKEN: ${{ secrets.HF_TOKEN }}
        run: |
          git config --global user.email "actions@github.com"
          git config --global user.name "GitHub Actions"
          git clone https://huggingface.co/spaces/your-hf-username/your-hf-space-name frontend_space
          cp tourism_project/deployment/app.py frontend_space/
          cp tourism_project/deployment/rf_model.joblib frontend_space/
          cp tourism_project/deployment/requirements.txt frontend_space/
          cp tourism_project/deployment/encoder.joblib frontend_space/
          cp tourism_project/data/tourism.csv frontend_space/
          cd frontend_space
          git add .
          git commit -m "Update Streamlit app and model"
          git push
